include/dlaf/eigensolver/tridiag_solver/rot.h

```diff
@@ -0,0 +1,146 @@
+          const LocalTileIndex loc_tile{dist_sub.nextLocalTileFromGlobalElement<Coord::Row>(0),
+                                        dist_sub.nextLocalTileFromGlobalElement<Coord::Col>(rot.i)};
+          const std::size_t idx_x = to_sizet(dist_sub.localTileLinearIndex(loc_tile));
+          return tiles[idx_x];
+        }
+        return tile_ws;
+      }();
+      const auto& tile_y = [&]() -> matrix::Tile<T, Device::CPU> const& {
+        if (hasY) {
+          const LocalTileIndex loc_tile{dist_sub.nextLocalTileFromGlobalElement<Coord::Row>(0),
+                                        dist_sub.nextLocalTileFromGlobalElement<Coord::Col>(rot.j)};
+          const std::size_t idx_y = to_sizet(dist_sub.localTileLinearIndex(loc_tile));
+          return tiles[idx_y];
+        }
+        return tile_ws;
+      }();
+
+      const bool hasBothXY = hasX && hasY;
+
+      const TileElementIndex idx_x(0, rot.i % mb);  // TODO check if dist can be used
+      const TileElementIndex idx_y(0, rot.j % mb);
+
+      std::vector<ex::unique_any_sender<>> cps;
+      if (!hasBothXY) {
+        // TODO compute TAG
+
+        const comm::IndexT_MPI rank_partner = hasX ? rankColY : rankColX;
+
+        // TODO possible optimization, check if it is zero or not
+
+        auto sender = [idx = hasX ? idx_x : idx_y, n](comm::Communicator& comm,
+                                                      comm::IndexT_MPI rank_dest, comm::IndexT_MPI tag,
+                                                      const auto& tile, MPI_Request* req) {
+          DLAF_MPI_CHECK_ERROR(MPI_Isend(tile.ptr(idx), static_cast<int>(n),
+                                         dlaf::comm::mpi_datatype<T>::type, rank_dest, tag, comm, req));
+        };
+        auto cp_send =
+            dlaf::internal::whenAllLift(comm_row, rank_partner, 0, std::cref(hasX ? tile_x : tile_y)) |
+            dlaf::comm::internal::transformMPI(sender);
+
+        auto receiver = [n](comm::Communicator& comm, comm::IndexT_MPI rank_dest, comm::IndexT_MPI tag,
+                            const matrix::Tile<T, Device::CPU>& tile, MPI_Request* req) {
+          DLAF_MPI_CHECK_ERROR(MPI_Irecv(tile.ptr({0, 0}), static_cast<int>(n),
+                                         dlaf::comm::mpi_datatype<T>::type, rank_dest, tag, comm, req));
+        };
+        auto cp_recv = dlaf::internal::whenAllLift(comm_row, rank_partner, 0, std::cref(tile_ws)) |
+                       dlaf::comm::internal::transformMPI(receiver);
+        cps.emplace_back(std::move(cp_send));
+        cps.emplace_back(std::move(cp_recv));
+      }
+
+      // each one computes his own, but just stores either x or y
+      // (or both if are on the same rank)
+      T* x = hasX ? tile_x.ptr(idx_x) : tile_ws.ptr({0, 0});
+      T* y = hasY ? tile_y.ptr(idx_y) : tile_ws.ptr({0, 0});
+
+      tt::sync_wait(ex::when_all_vector(std::move(cps)) |
+                    ex::then([rot, n, x, y]() { blas::rot(n, x, 1, y, 1, rot.c, rot.s); }));
```

Comment:
Could you briefly explain the need for a `sync_wait` here (as we generally don't need it)?

Reply:
I'm not fully sure about this, so thanks for the question that gives me the chance to discuss about it.

Everything happens inside a task, which is given a list of rotations and more relevantly a list of tiles (the one inside the square sub-matrix selection).

Tiles lifetime is ensured till the end of the task. If we `start_detached` the sub-task above, afaik, it can be that the main task finishes releasing the tile, and the sub-task  might operate on a tile that no one is keeping alive anymore.

With `sync_wait`, we ensure that the task will wait for the conclusion of the sub-task. This has a "side-effect" of serialising the loop over rotations, which, afaict, it is what happens also in the current local implementation.

It must be said that it might also not be the best implementation ever (surely it isn't).

As discussed with @rasolca offline this morning, there are other options.
For instance, an option would be to schedule all communications and just later schedule computations (still waiting on completion of needed communications); this would require to allocate more workspaces (currently it is just one, and it represents another limiting factor to the parallelisation, but matches the one imposed by `sync_wait`). 
Another variant would be to avoid allocating one workspaces for each rotation and just allocate a bunch of them and group rotations together and execute them in chunks.
Somehow orthogonally to this, we can also opt to pack them to reduce the number of communications happening.

Anyway, even just considering this basic implementation, I'm really looking forward to your comments in order to understand if I'm missing something and it can be done differently and better. Thanks in advance!

Reply:
I think this is probably worth an offline discussion so I'll just briefly respond: `let_value` would probably work as a drop-in replacement of the `transform` and you'd get the tiles to live long enough. Then you'd instead return the sender instead of `sync_wait`ing it as the last thing in the function. What might change the discussion is if you have other plans for the `transform` (that creates a task with `givens_rots_fn`), since at the moment it's not necessary. I'm guessing you're planning to implement a GPU version at some point and then the `transform` will bactually be required. But even then perhaps the `transform` can be called inside `givens_rots_fn`. Slightly too many unknowns still at this point to be able to say anything more.

Reply:
@msimberg I added some documentation trying to explain why I think `sync_wait` is currently needed with the current implementation.

We can discuss about it offline if you want to have a support in getting the idea or if you want to share some suggestions. 😉

Reply:
Thanks! Yes, we can discuss details offline as I'm almost certain this can be rewritten using `let_value`.